proposals/ref-struct-interfaces.md

```diff
@@ -35,6 +35,34 @@ The ability to implement interfaces is only useful when combined with the abilit
 
 To allow for interfaces to cover the full expressiveness of a `ref struct` and the lifetime issues they can present, the language will allow `[UnscopedRef]` to appear on interface methods and properties. When a `struct` / `ref struct` member implements an interface member with a `[UnscopedRef]` attribute, that member must also be decorated with `[UnscopedRef]`. The attribute is ignored when a `class` implements the interface.
 
+```csharp
+interface I1
+{
+    [UnscopedRef]
+    ref int Prop { get; }
+}
+
+struct S1 : I1
+{
+    // Error: missing [UnscopedRef]
```

Comment:
Please read the whole comment if you're not convinced, I explain my logic from multiple angles, including one based on already existing scoped rules in C# (linked).

I don't see why this wouldn't be allowed, consider that I can always safely work around this arbitrary limitation in the case of properties:
```csharp
struct S1 : I1
{
    ref int Prop { get; }
    [UnscopedRef] ref int I1.Prop => ref Prop;
}
```
By using an explicit interface implementation.

To me, this indicates that the implementation should not require `[UnscopedRef]` to be on the member, even if it's on the interface declaration, since a non-`[UnscopedRef]` implementation is always a valid `[UnscopedRef]` implementation.

This is because, `[UnscopedRef]` on a struct member (not the parameter, the member itself) simply indicates that the return value may include a ref which overlaps with `ref this` (or via `RefStruct&` with applicable annotations), it does not necessitate that it does. That's also why you cannot have an `[UnscopedRef]` member in a class, since it can never return a ref which overlaps with `ref this` (unless it's passed in separately obviously).

In fact, if you pretend that we pass this as an actual parameter to structs, a normal method `ref int X() => ...;` is effectively `ref int X(scoped ref ThisType this) => ...;`, whereas an unscoped-ref method `[UnscopedRef] ref int Y() => ...;` is effectively `ref int Y(ref ThisType this) => ...;`. If you look at the [delegate conversion rules](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-11.0/low-level-struct-improvements#parameter-scope-variance) around `scoped`, you'll notice that you are allowed to add `scoped` to a `ref` parameter in an override. Therefore, if I'm expected to implement `ref int Y(ref ThisType this) => ...;` (`[UnscopedRef]`), then I can implement it with `ref int Y(scoped ref ThisType this) => ...;` instead if I want.

The first step of my logic is explained [here](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-11.0/low-level-struct-improvements#provide-unscoped) (that is, how it impacts the implicit `ref this` parameter). And the second is explained on the delegate conversion & overriding rules linked above.

> When applying `[UnscopedRef]` to an instance method of a `struct` it has the impact of modifying the implicit `this` parameter. This means `this` acts as an unannotated `ref` of the same type.

I have come up with all permutations of parameters to show that you can always implement a `[UnscopedRef]` member with a non-`UnscopedRef` member, no matter what parameters a method has, and as shown earlier it also works for properties: [link](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBLANgHwAEAmARgFgAoQgBgAJDSA6AEWwEMBzAOwgGcM2MHyYBhCABMYAQW7tcATz7Y+Abiq0GzAEoBXboIC2MMREMAHPDCgBlawDchMNVQ0BmOgKi6wGOjYpKAG8qOjC6WAAzOmwDOmlSOgBeAD4ImGiAVW4+dkiTADldXFxtDIAeWIwUgAoASnVKcLoAbWy+SHMYCTLIgF106Kr44mS0qPjSRtDwieGAIUDmmpnmsLm49jRPTu7BmLjgbdiDvzBtjoguiQPTugk0VbWIXT9hmG22nN2ejIGXt5xSLbGzmdjcSoGNKcC4/fxgiFVNIAC0eTTW6wy8PBkOqMVhVz2E1BOKRdAAVtsAdjEVC6ABrT7tH69f6vGm4tK4J51J4hdEYmDJe4ZdjFDCNDFhaJJEWRMW4CVPZr04VSeXiyVS3Bq0Wa5XhQgAdn27TyhWKpQqSPqWrCAF9XALWszCb9+vsFsQnitnc0Nn4tjs3Z7DsduHdzsHrrdhg8DWFqe8md83ay6EmgSCEZy6DDo3sSbS8aiE/si7nsASY8Sc2TKRn2RWyYyXanrunqc26dznXUxvtFjUJkGJkdblGTg9G34PjO6MC89tUZ7thMG9TW9TcA0nc1u3jREtwr6paHA9WiVjhuOTsMo5cY3e4vG/eFM7OU4/up32cNFwe0KXjcgF0KWb6YtEoFVgWNy1qSdIbk2dZ0q2Xzfu6bJ+KBvbNLyzr8meQqyuqCpKhBC66hqip2msqokXqNFljqDHUeRZ7GnKZG0Y6lBPOhLJ/ByZKiN6zqnlKAZ0EGGHnnQt4RvewGxi+aJnh+dBzgJaZCRpAEoXi+ayaB4FnvBxZpDBsnmbmSHYQZaRoa6Ha6chCF4rh4T9qkdBHsOWKjliCl0JOEbTtSc7Uou+YrnMa5YnZDJUuyO7TM6+i5Pk5YOXQLDHmEEkYlJMlwlJwVKbBKl+K+6l/nEWnOT+rmAn4+nuUBlUmWpklYtByk2fWyX2e1SVthhv7DRZdCeWE+HNIRUrEVx+oUTKy1MRR9Hrex2pUdxZacRMABUNSAUddQ0Dxe7hNpLkegNdIsGJyxlsV/XXmGVUhcpz7Vd1GIaQ17ZNR6enZiNRlwl1r29TlVmlbDI2JaBTnA5h844TyA4THl/nRIF0TBaF9xDZppPRcu8VDFTFKk1uKW7pQvFAA=).

But you would definitely not want the implementation of a non-`[UnscopedRef]` interface member to have `[UnscopedRef]`, since that would break the contract (as is the case currently).

Reply:
> I have come up with all permutations of parameters to show that you can always implement ...

The permutations example isn't persuasive to me. Had cases in the past where did permutation arguments but found later that we missed permutations / interactions with other features. Lifetime interactions are very complex. An argument on the underlying lifetime rules would be more persuasive. 

> Please read the whole comment if you're not convinced ... 

One point you haven't addressed is what is the value in removing this requirement? The value in keeping is that it's a simple rule, no need to prove the absence of it is safe, etc ... What is the value in going through that exercise?

Reply:
> The permutations example isn't persuasive to me. Had cases in the past where did permutation arguments but found later that we missed permutations / interactions with other features. Lifetime interactions are very complex. An argument on the underlying lifetime rules would be more persuasive.

Which I also presented (repeated more clearly below).

It would effectively be a change to existing rules to disallow it. The existing rules explain how to interpret `[UnscopedRef]` members on structs: 

From https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-11.0/low-level-struct-improvements#provide-unscoped

> When applying `[UnscopedRef]` to an instance method of a `struct` it has the impact of modifying the implicit `this` parameter. This means `this` acts as an unannotated `ref` of the same type.

By default (without `[UnscopedRef]`), `this` is passed as `scoped ref this`, but with `[UnscopedRef]` on the method, it's passed as `ref this`.

From https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-11.0/low-level-struct-improvements#parameter-scope-variance

> The `scoped` modifier and `[UnscopedRef]` attribute (see below) on parameters also impacts our object overriding, interface implementation and `delegate` conversion rules. The signature for an override, interface implementation or `delegate` conversion can:
- Add `scoped` to a `ref` or `in` parameter
- ...

Note these rules apply to interface implementation, which is what we're discussing here. It says we can add `scoped` to a `ref` parameter, meaning that we can implement `ref int M(ref this)` (`[UnscopedRef] ref int M()`) with `ref int M(scoped ref this)` (`ref int M()`). To change this would effectively mean changing the existing rules on lifetimes.

Furthermore, to prove its safety more directly, since `ref int M()` is effecively `ref int M(scoped ref this)`, and `[UnscopedRef] ref int M()` is effecively `ref int M(ref this)` - the `scoped` in `ref int M(scoped ref this)` means that `ref this` will not ever escape; this means that it is compatible with `ref int M(ref this)` where `ref this` is allowed to escape. i.e., `[UnscopedRef] ref int M()` requires that `ref this` is considered to be potentially stored in any `ref` after the method exits, whereas `ref int M()` guarantees that it won't be - if something won't ever happen (`ref this` escaping - no `[UnscopedRef]`), it is safe to treat it as a possibility (`[UnscopedRef]`), it just means that there are other safe operations which wouldn't be allowed.

> One point you haven't addressed is what is the value in removing this requirement? The value in keeping is that it's a simple rule, no need to prove the absence of it is safe, etc ... What is the value in going through that exercise?

Consider that I have 2 implementations for something, which are exposed through structs. One has the fields directly on itself, whereas the other one has some of the fields on helper (class) fields. If I want to expose an interface for retrieving one of these fields by reference, then I need to expose it as `[UnscopedRef]`, but only the implementation which has the fields on the struct directly needs to actually use `[UnscopedRef]` on its public member, the other one can just do it as a normal ref.

To mark the other one as `[UnscopedRef]` would potentially (in my case, actually) unnecessarily reduce the possible ways it could be used. Having to define an explicit implementation is something I can do (and will if I have to), but it also should be unnecessary since it's safe as I showed above with the current rules.

This is not an academic example, this is something I have in (non-exposed) code I actually have already written (just without the safety of being able to use `[UnscopedRef]` on the interface member definition currently). I would like to gain that safety for obvious reasons - no need to be unsafe when there's not supposed to actually be anything unsafe about the code.

e.g., what I do today:
```csharp
interface I1
{
    ref int PotentialField { get; }
}

struct S1 : I1
{
    int field;
    [UnscopedRef] public ref int PotentialField => ref field;
    unsafe ref int I1.PotentialField => ref PotentialField; //ignore warning
}

struct S2 : I1
{
    SomeClass field;
    public int PotentialField => ref field.fieldOfTypeInt; //can implement it here
}
```

Ideally I could just write:
```csharp
interface I1
{
    [UnscopedRef] ref int PotentialField { get; }
}

struct S1 : I1
{
    int field;
    [UnscopedRef] public ref int PotentialField => ref field;
}

struct S2 : I1
{
    SomeClass field;
    public int PotentialField => ref field.fieldOfTypeInt; //can implement it here
}
```

Reply:
> Had cases in the past where did permutation arguments but found later that we missed permutations / interactions with other features

Out of curiosity - what was an example of this :)

Reply:
Thought about it some more this morning and decided I agreed that it's safe. I added a rationale at the bottom of the spec. Essentially, though the approach I took is make `this` explicit parameter and verify that _effectively adding / removing `scoped` in the ways that `[UnscopedRef]` work here is already legal for parameters so this is already established.

Reply:
> Out of curiosity - what was an example of this :)

There were a number of them. Probably the one I most remember, and was missed in the initial ref reassignment implementation in 7.3, is the following: 

```csharp
Span<int> stackSpan = stackalloc int[42];
Span<int> heapSpan = new int[42];

ref Span<int> refStackSpan = ref stackSpan;
ref Span<int> refHeapSpan = ref heapSpan;

// Must be an error
refStackSpan = ref refHeapSpan;

// If it's allowed then you can do this which writes a stack value to 
// heapSpan ... oops
refStackSpan = new stackalloc int[13];
```

This is a case where we went through the permutations and decided it was legal. While we succeeded in enumerating the permutations, we failed in understanding their implications. We errored in saying that it's always safe to make the lifetime smaller as _seems_ to occur here. In reality it's only safe to make the lifetime smaller in some cases but not others. 

The [annotations][1] part of the spec does a cleaner job of expressing this type of problem as lifetimes are modeled with variance. In most cases lifetime is covariant (can be made more specific / smaller). When visualizing `ref` as a type where lifetime is a parameter (as it is on other types) it becomes clear that lifetime there must be invariant. Otherwise it's easy to violate lifetimes. A few other cases like this were found when we modeled in this fashion. 

[1]: https://github.com/dotnet/csharplang/blob/main/proposals/csharp-11.0/low-level-struct-improvements.md#annotations